File: packages/js/src/extensions/security.ts
/**
 * Security Extensions for JSON-LD
 *
 * Provides:
 * - @integrity: SHA-256/384/512 hash verification for contexts
 * - Context allowlists: Restrict which remote contexts can be loaded
 * - Resource limits: Prevent DoS via recursion/exhaustion
 */

import * as crypto from 'crypto';
import { KEYWORD_INTEGRITY } from '../keywords';
import { IntegrityContext, ResourceLimits, ContextAllowlist } from '../types';

// ── Default Resource Limits ───────────────────────────────────────

export const DEFAULT_RESOURCE_LIMITS: Required<ResourceLimits> = {
  maxContextDepth: 10,
  maxGraphDepth: 100,
  maxDocumentSize: 10 * 1024 * 1024, // 10 MB
  maxExpansionTime: 30_000, // 30 seconds
};

// ── Context Integrity ─────────────────────────────────────────────

type HashAlgorithm = 'sha256' | 'sha384' | 'sha512';

/**
 * Computes an integrity hash for a context string/object.
 *
 * @param context - The context content (string or object to be serialized)
 * @param algorithm - Hash algorithm (default: sha256)
 * @returns Integrity string in format "algorithm-base64hash"
 */
export function computeIntegrity(
  context: string | object,
  algorithm: HashAlgorithm = 'sha256'
): string {
  const content = typeof context === 'string'
    ? context
    : JSON.stringify(context);

  const nodeAlg = algorithm === 'sha256' ? 'sha256'
    : algorithm === 'sha384' ? 'sha384'
    : 'sha512';

  const hash = crypto.createHash(nodeAlg).update(content, 'utf8').digest('base64');
  return `${algorithm}-${hash}`;
}

/**
 * Verifies a context's content against its declared integrity hash.
 *
 * @param context - The context content
 * @param declaredIntegrity - The expected integrity string
 * @returns true if the hash matches
 */
export function verifyIntegrity(
  context: string | object,
  declaredIntegrity: string
): boolean {
  const [algorithm, _expectedHash] = declaredIntegrity.split('-', 2) as [string, string];

  if (!['sha256', 'sha384', 'sha512'].includes(algorithm)) {
    throw new Error(`Unsupported hash algorithm: ${algorithm}`);
  }

  const computed = computeIntegrity(context, algorithm as HashAlgorithm);
  return computed === declaredIntegrity;
}

/**
 * Creates a context reference with integrity verification.
 */
export function integrityContext(
  contextUrl: string,
  contextContent: string | object,
  algorithm: HashAlgorithm = 'sha256'
): IntegrityContext {
  return {
    '@id': contextUrl,
    '@integrity': computeIntegrity(contextContent, algorithm),
  };
}

// ── Context Allowlist ─────────────────────────────────────────────

/**
 * Checks whether a context URL is permitted by the allowlist configuration.
 */
export function isContextAllowed(
  contextUrl: string,
  config: ContextAllowlist
): boolean {
  if (config.blockRemoteContexts) {
    return false;
  }

  // Check exact matches
  if (config.allowed?.includes(contextUrl)) {
    return true;
  }

  // Check patterns
  if (config.patterns) {
    for (const pattern of config.patterns) {
      if (typeof pattern === 'string') {
        // Simple glob: convert * to regex
        const regex = new RegExp(
          '^' + pattern.replace(/\*/g, '.*').replace(/\?/g, '.') + '$'
        );
        if (regex.test(contextUrl)) return true;
      } else if (pattern instanceof RegExp) {
        if (pattern.test(contextUrl)) return true;
      }
    }
  }

  // If allowlist is configured but URL didn't match, deny
  if ((config.allowed && config.allowed.length > 0) ||
      (config.patterns && config.patterns.length > 0)) {
    return false;
  }

  // No allowlist configured = allow all
  return true;
}

/**
 * Creates a secure document loader that enforces allowlists and integrity.
 *
 * Wraps an existing jsonld.js document loader with security checks.
 *
 * @param baseLoader - The underlying document loader (from jsonld.js)
 * @param config - Allowlist configuration
 * @param integrityMap - Optional map of context URL → expected integrity hash
 */
export function createSecureDocumentLoader(
  baseLoader: (url: string) => Promise<any>,
  config: ContextAllowlist,
  integrityMap?: Map<string, string>
): (url: string) => Promise<any> {
  return async (url: string) => {
    // Enforce allowlist
    if (!isContextAllowed(url, config)) {
      throw new Error(
        `Context URL blocked by allowlist: ${url}. ` +
        `Allowed: ${JSON.stringify(config.allowed ?? [])}`
      );
    }

    // Load the document
    const result = await baseLoader(url);

    // Verify integrity if declared
    if (integrityMap?.has(url)) {
      const expected = integrityMap.get(url)!;
      const content = typeof result.document === 'string'
        ? result.document
        : JSON.stringify(result.document);

      if (!verifyIntegrity(content, expected)) {
        throw new Error(
          `Context integrity verification failed for ${url}. ` +
          `Expected: ${expected}`
        );
      }
    }

    return result;
  };
}

// ── Resource Limits ───────────────────────────────────────────────

/**
 * Validates a document against resource limits before processing.
 *
 * @throws Error if any limit is exceeded
 */
export function enforceResourceLimits(
  document: string | object,
  limits: ResourceLimits = DEFAULT_RESOURCE_LIMITS
): void {
  const resolved = { ...DEFAULT_RESOURCE_LIMITS, ...limits };

  // Check document size
  const content = typeof document === 'string'
    ? document
    : JSON.stringify(document);

  if (content.length > resolved.maxDocumentSize) {
    throw new Error(
      `Document size ${content.length} bytes exceeds limit of ${resolved.maxDocumentSize} bytes`
    );
  }

  // Check graph depth
  const parsed = typeof document === 'string' ? JSON.parse(document) : document;
  const depth = measureDepth(parsed);

  if (depth > resolved.maxGraphDepth) {
    throw new Error(
      `Document nesting depth ${depth} exceeds limit of ${resolved.maxGraphDepth}`
    );
  }
}

/**
 * Creates a processing timeout wrapper.
 */
export function withTimeout<T>(
  promise: Promise<T>,
  timeoutMs: number,
  operation: string = 'processing'
): Promise<T> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`${operation} timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    promise
      .then((result) => {
        clearTimeout(timer);
        resolve(result);
      })
      .catch((err) => {
        clearTimeout(timer);
        reject(err);
      });
  });
}

// ── Helpers ───────────────────────────────────────────────────────

function measureDepth(obj: any, current: number = 0): number {
  if (obj == null || typeof obj !== 'object') return current;

  let maxDepth = current;

  if (Array.isArray(obj)) {
    for (const item of obj) {
      maxDepth = Math.max(maxDepth, measureDepth(item, current + 1));
    }
  } else {
    for (const value of Object.values(obj)) {
      maxDepth = Math.max(maxDepth, measureDepth(value, current + 1));
    }
  }

  return maxDepth;
}

File: packages/python/src/jsonld_ex/security.py
"""Security Extensions for JSON-LD."""

from __future__ import annotations
import hashlib
import base64
import json
import re
from typing import Any, Optional


DEFAULT_RESOURCE_LIMITS = {
    "max_context_depth": 10,
    "max_graph_depth": 100,
    "max_document_size": 10 * 1024 * 1024,  # 10 MB
    "max_expansion_time": 30,  # seconds
}

SUPPORTED_ALGORITHMS = ("sha256", "sha384", "sha512")


def compute_integrity(
    context: str | dict | Any, algorithm: str = "sha256"
) -> str:
    """Compute an integrity hash for a context."""
    if context is None:
        raise TypeError("Context must not be None")
    if algorithm not in SUPPORTED_ALGORITHMS:
        raise ValueError(f"Unsupported algorithm: {algorithm}")
    try:
        content = context if isinstance(context, str) else json.dumps(context, sort_keys=True)
    except (TypeError, ValueError) as exc:
        raise TypeError(f"Context is not JSON-serializable: {exc}") from exc
    h = hashlib.new(algorithm, content.encode("utf-8")).digest()
    b64 = base64.b64encode(h).decode("ascii")
    return f"{algorithm}-{b64}"


def verify_integrity(context: str | dict | Any, declared: str) -> bool:
    """Verify context content against its declared integrity hash."""
    parts = declared.split("-", 1)
    if len(parts) != 2 or parts[0] not in SUPPORTED_ALGORITHMS:
        raise ValueError(f"Invalid integrity string: {declared}")
    computed = compute_integrity(context, parts[0])
    return computed == declared


def integrity_context(
    url: str, content: str | dict | Any, algorithm: str = "sha256"
) -> dict[str, str]:
    """Create a context reference with integrity verification."""
    return {"@id": url, "@integrity": compute_integrity(content, algorithm)}


def is_context_allowed(url: str, config: dict[str, Any]) -> bool:
    """Check if a context URL is permitted by an allowlist configuration."""
    if config.get("block_remote_contexts", False):
        return False
    allowed = config.get("allowed", [])
    if url in allowed:
        return True
    for pattern in config.get("patterns", []):
        if isinstance(pattern, str):
            regex = "^" + re.escape(pattern).replace(r"\*", ".*").replace(r"\?", ".") + "$"
            if re.match(regex, url):
                return True
    if allowed or config.get("patterns"):
        return False
    return True


_MAX_RECURSION_DEPTH = 500  # Safety cap for _measure_depth


def enforce_resource_limits(
    document: str | dict | Any,
    limits: Optional[dict[str, int]] = None,
) -> None:
    """Validate document against resource limits before processing."""
    if document is None:
        raise TypeError("Document must not be None")
    resolved = {**DEFAULT_RESOURCE_LIMITS, **(limits or {})}
    if isinstance(document, str):
        content = document
        if len(content) > resolved["max_document_size"]:
            raise ValueError(
                f"Document size {len(content)} exceeds limit {resolved['max_document_size']}"
            )
        try:
            parsed = json.loads(content)
        except json.JSONDecodeError as exc:
            raise ValueError(f"Document is not valid JSON: {exc}") from exc
    elif isinstance(document, (dict, list)):
        try:
            content = json.dumps(document)
        except (TypeError, ValueError) as exc:
            raise TypeError(f"Document is not JSON-serializable: {exc}") from exc
        if len(content) > resolved["max_document_size"]:
            raise ValueError(
                f"Document size {len(content)} exceeds limit {resolved['max_document_size']}"
            )
        parsed = document
    else:
        raise TypeError(f"Document must be a str, dict, or list, got: {type(document).__name__}")
    depth = _measure_depth(parsed)
    if depth > resolved["max_graph_depth"]:
        raise ValueError(
            f"Document depth {depth} exceeds limit {resolved['max_graph_depth']}"
        )


def _measure_depth(obj: Any, current: int = 0) -> int:
    if current > _MAX_RECURSION_DEPTH:
        return current  # Safety cap to prevent stack overflow
    if obj is None or not isinstance(obj, (dict, list)):
        return current
    max_depth = current
    items = obj if isinstance(obj, list) else obj.values()
    for item in items:
        max_depth = max(max_depth, _measure_depth(item, current + 1))
    return max_depth
